BinaryTree.java

```diff
@@ -0,0 +1,137 @@
+import java.util.InputMismatchException;
+
+/**
+ * Created by basescu on 29.05.17.
+ */
+public class BinaryTree {
+    private Integer info;
+    private BinaryTree left;
+    private BinaryTree right;
+
+    public BinaryTree(int info, BinaryTree left, BinaryTree right) {
+        this.info = info;
+        this.left = left;
+        this.right = right;
+    }
+
+    public Integer getInfo() {
+        return info;
+    }
+
+    public String printAncestors(Integer key) {
+        return this.printAncestors(key, "");
+    }
+
+    private String printAncestors(Integer key, String ancestors) {
+        if(this.info == key)
+            return ancestors;
+        if (this.left == null && this.right == null)
+            return "";
+        StringBuilder sb = new StringBuilder(ancestors.length() + 10);
+        sb.append(this.info);
+        sb.append(" ");
+        sb.append(ancestors);
+        String left = "";
+        String right = "";
+
+        if(this.left != null) {
+           left = this.left.printAncestors(key, sb.toString());
+        }
+        if(this.right != null) {
+            right = this.right.printAncestors(key, sb.toString());
+        }
+
+        if(left != "")
+            return left;
+        return right;
+    }
+
+    private BinaryTree lowestCommonAncestor(BinaryTree tree1, BinaryTree tree2) {
+        if(tree1 == null || tree2 == null)
+            return null;
+
+        BinaryTree leftRes = null;
+        BinaryTree rightRes = null;
+
+        if (this.left != null)
+            leftRes = this.left.lowestCommonAncestor(tree1, tree2);
+        if (this.right != null)
+            rightRes = this.right.lowestCommonAncestor(tree1, tree2);
+        if(this.left == null)
+            return rightRes;
+        if(this.right == null)
+            return leftRes;
+
+        if(this.left.contains(tree1) && this.right.contains(tree2))
+            return this;
+        if(this.left.contains(tree2) && this.right.contains(tree1))
+            return this;
+
+        if(leftRes != null)
+            return leftRes;
+        return rightRes;
+    }
+
+    private boolean contains(BinaryTree contained) {
+        if(this.info == contained.info) {
+           if (this.isIdentical(contained))
+               return true;
+        }
+
+        boolean containedLeft = false;
+        boolean containedRight = false;
+
+        if(this.left != null)
+            containedLeft = this.left.contains(contained);
+
+        if(this.right != null)
+            containedRight = this.right.contains(contained);
+
+        return (containedLeft || containedRight);
+    }
+
+    private boolean isIdentical(BinaryTree tree) {
```

Comment:
I thought we wanted to find the ancestor of two subtrees: for instance because keys are not unique, but they're unique if we provide the entire subtree. I changed it to keys in the new code version, assuming keys are unique.